Branch-SDK/src/main/java/io/branch/referral/NativeShareLinkManager.java

```diff
@@ -0,0 +1,118 @@
+        if (INSTANCE == null) {
+            // synchronize the block to ensure only one thread can execute at a time
+            synchronized (NativeShareLinkManager.class) {
+                // check again if the instance is already created
+                if (INSTANCE == null) {
+                    INSTANCE = new NativeShareLinkManager();
+                }
+            }
+        }
+        return INSTANCE;
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP_MR1)
+    void shareLink(@NonNull Activity activity, @NonNull BranchUniversalObject buo, @NonNull LinkProperties linkProperties, @Nullable Branch.BranchNativeLinkShareListener callback, String title, String subject) {
+
+        nativeLinkShareListenerCallback_ = new NativeLinkShareListenerWrapper(callback, linkProperties, buo);
+
+        try {
+            buo.generateShortUrl(activity, linkProperties, new Branch.BranchLinkCreateListener() {
+                @Override
+                public void onLinkCreate(String url, BranchError error) {
+                    if (error == null) {
+                        SharingUtil.share(url, title, subject, activity);
+                    } else {
+
+                        if (callback != null) {
+                            callback.onLinkShareResponse(url, error);
+                        } else {
+                            BranchLogger.v("Unable to share link " + error.getMessage());
+                        }
+                        if (error.getErrorCode() == BranchError.ERR_BRANCH_NO_CONNECTIVITY
+                                || error.getErrorCode() == BranchError.ERR_BRANCH_TRACKING_DISABLED) {
+                            SharingUtil.share(url, title, subject, activity);
+                        }
+                    }
+                }
+            });
+
+        } catch (Exception e) {
+            e.printStackTrace();
+            if (nativeLinkShareListenerCallback_ != null) {
+                nativeLinkShareListenerCallback_.onLinkShareResponse(null, new BranchError("Trouble sharing link", BranchError.ERR_BRANCH_NO_SHARE_OPTION));
+            } else {
+                BranchLogger.v("Unable to share link. " + e.getMessage());
+            }
+        }
+    }
+
+    public Branch.BranchNativeLinkShareListener getLinkShareListenerCallback() {
+        return nativeLinkShareListenerCallback_;
+    }
+
+    /**
+     * Class for intercepting share sheet events to report auto events on BUO
+     */
+    private class NativeLinkShareListenerWrapper implements Branch.BranchNativeLinkShareListener {
+        private final Branch.BranchNativeLinkShareListener originalCallback_;
+        private final BranchUniversalObject buo_;
+        private String channelSelected_;
+
+        NativeLinkShareListenerWrapper(Branch.BranchNativeLinkShareListener originalCallback, LinkProperties linkProperties, BranchUniversalObject buo) {
+            originalCallback_ = originalCallback;
```

Comment:
I'd name as just `branchNativeLinkShareListener`, maybe my own reading difficulty but `originalCallback_` makes me think one was duplicated elsewhere.

Reply:
Its a wrapper API, so in that context word original was used. As its confusing, I am updating it to the one you have suggested.